File: jtaf-web/src/java/ch/jtaf/entity/Category.java
package ch.jtaf.entity;

import java.util.ArrayList;
import java.util.List;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import javax.persistence.ManyToOne;
import javax.persistence.NamedQueries;
import javax.persistence.NamedQuery;
import javax.persistence.OneToMany;
import javax.persistence.OrderColumn;

@Entity
@NamedQueries({
    @NamedQuery(name = "Category.findAll", query = "select c from Category c where c.series_id = :series_id order by c.abbreviation"),
    @NamedQuery(name = "Category.findBySeriesAndYearAndGender", query = "select c from Category c where c.series_id = :series_id and c.gender = :gender and :year between c.yearFrom and c.yearTo")
})
public class Category {

    @Id
    @GeneratedValue
    private Long id;
    private String abbreviation;
    private String name;
    private int yearFrom;
    private int yearTo;
    private String gender;
    private Long series_id;
    @OneToMany
    @OrderColumn(name = "position")
    private List<Event> events = new ArrayList<>();

    public List<Event> getEvents() {
        return events;
    }

    public void setEvents(List<Event> events) {
        this.events = events;
    }

    public String getAbbreviation() {
        return abbreviation;
    }

    public void setAbbreviation(String abbreviation) {
        this.abbreviation = abbreviation;
    }

    public int getYearFrom() {
        return yearFrom;
    }

    public void setYearFrom(int yearFrom) {
        this.yearFrom = yearFrom;
    }

    public int getYearTo() {
        return yearTo;
    }

    public void setYearTo(int yearTo) {
        this.yearTo = yearTo;
    }

    public String getGender() {
        return gender;
    }

    public void setGender(String gender) {
        this.gender = gender;
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Long getSeries_id() {
        return series_id;
    }

    public void setSeries_id(Long series_id) {
        this.series_id = series_id;
    }

    @Override
    public int hashCode() {
        int hash = 5;
        hash = 19 * hash + (this.id != null ? this.id.hashCode() : 0);
        return hash;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == null) {
            return false;
        }
        if (getClass() != obj.getClass()) {
            return false;
        }
        final Category other = (Category) obj;
        if (this.id != other.id && (this.id == null || !this.id.equals(other.id))) {
            return false;
        }
        return true;
    }
}

File: src/main/kotlin/ch/jtaf/entity/Category.kt
package ch.jtaf.entity

import ch.jtaf.entity.Gender.MALE
import javax.persistence.*
import javax.persistence.GenerationType.IDENTITY

@Entity
data class Category(
        @Id @GeneratedValue(strategy = IDENTITY)
        var id: Long? = null,

        var abbreviation: String = "",
        var name: String = "",
        var yearFrom: Int = 1900,
        var yearTo: Int = 9999,
        var gender: Gender = MALE,

        var seriesId: Long? = null
) {
    @OneToMany
    @OrderColumn(name = "position")
    @JoinTable(name = "category_event",
            joinColumns = [(JoinColumn(name = "category_id", referencedColumnName = "id"))],
            inverseJoinColumns = [(JoinColumn(name = "event_id", referencedColumnName = "id"))])
    var events: MutableList<Event> = ArrayList()

    @OneToMany
    @JoinTable(name = "category_athlete",
            joinColumns = [(JoinColumn(name = "category_id", referencedColumnName = "id"))],
            inverseJoinColumns = [(JoinColumn(name = "athlete_id", referencedColumnName = "id"))])
    var athletes: MutableList<Athlete> = ArrayList()
}